### ISSUE TITLE
Add support for using lazyStyleTag use() from a JavaScript module (injecting into multiple shadow DOMs)

### ISSUE THREAD

**Issue by mikeaustin** [webpack-contrib/style-loader#565]
### Feature Proposal

When using `lazyStyleTag` and `use({ target: shadowRoot })` inside a shared JavaScript module, only the first call to use injects a `style` element. The other shadow elements that call `use()` have no styles. I've duplicated the module and loaded each and the problem does not exist. There may be other options such somehow forcing separate module instances, but I'm not sure.

In index.js, I believe the following condition prevents use() from injecting duplicate styles:

```
if (!(refs++)) {
    update = API(content, options);
}
```

https://github.com/webpack-contrib/style-loader/blob/master/src/index.js#L131

One possible solution is to give an id to each injected `style` element, instead of having a global `refs` value.

### Feature Use Case

loading JS modules injected into shadow elements, so modules can be loaded dynamically.

### Please paste the results of `npx webpack-cli info` here, and mention other relevant information

  System:
    OS: macOS 12.1
    CPU: (4) x64 Intel(R) Core(TM) i5-8210Y CPU @ 1.60GHz
    Memory: 122.11 MB / 16.00 GB
  Binaries:
    Node: 14.18.1 - /usr/local/bin/node
    npm: 6.14.15 - /usr/local/bin/npm
  Browsers:
    Chrome: 101.0.4951.64
    Chrome Canary: 104.0.5071.0
    Firefox: 98.0.1
    Safari: 15.2
    Safari Technology Preview: 15.4
  Packages:
    css-loader: ^6.7.1 => 6.7.1 
    css-modules-typescript-loader: ^4.0.1 => 4.0.1 
    styles-loader: ^4.0.1 => 4.0.1 
    ts-loader: ^9.3.0 => 9.3.0 
    webpack-cli: ^4.9.2 => 4.9.2

**Comment by alexander-akait** [comment]
It it possible to create memory leaking and it will have very bad perf...

**Comment by mikeaustin** [comment]
HI @alexander-akait, can you elaborate on bad performance? I'm open to other suggestions. What about passing a key in options that can be used to associate the instance?

**Comment by mikeaustin** [comment]
Hi @alexander-akait, thank you for the reply. The original message covers the issue – loading a JS module dynamically, multiple times, only the first call to use() injects the style element. Since JS module instances are shared (loading a second time returns the original instance), only one style element is ever injected.
At the root of the issue, it’s because import(), static or dynamic, shares instances, and that instance uses the use() function. I don’t know if there’s a way to force a module to return a new instance. import() is used to load a React component dynamically. A function “mount” is exported in each module which passes the shadow root. The component then calls use() with that shadow root as target.

Thank you. Would another option to be to create a new instance manually? So instead of styles.use(), styles.create().

**Comment by alexander-akait** [comment]
@mikeaustin Can you create small example? I want to look how you use it to provide better solution

**Comment by andriichumak** [comment]
I'm having the same issue. It can be reproduced easily with the example from the docs: https://webpack.js.org/loaders/style-loader/#custom-elements-shadow-dom Just import the resulting js to some html file and create several instances of the custom element, like so:
```html
<custom-square></custom-square>
<custom-square></custom-square>
```
You'll notice that only the first element is styled. Would appreciate any advice on how to work around this properly. So far I'm thinking of some sort of MiniCssExtractPlugin + AssetsPlugin combination, but don't really like that option.

Ideally, there should be a `lazyLinkTag` injectType + an option to insert several links per runtime (one per Shadow Root).

**Comment by alexander-akait** [comment]
@andriichumak Do you keep styles in JS file or extract them for production?

**Comment by andriichumak** [comment]
I would prefer to have styles extracted, but it's not a hard requirement for my project. If we would be able to inject `style` tag into multiple places with `lazyStyleTag` - that would already be a big improvement.

Another option I can think of is to make the loader return the actual path to the generated css file instead of an empty object when using regular `styleTag`. This way I would be able to inject the `link` tags manually.

```javascript
import pathToFile from 'my.css';

// pathToFile = e.g. /assets/my.css
```

**Comment by alexander-akait** [comment]
@andriichumak Can you create small reproducible example to show usage?

**Comment by mditullio** [comment]
Had exactly the same issue today. It's a pity that in webpack there's no easy way to export a CSS module URL(s) to extracted files.

I'm considering reverting back and use css-loader exporting styles as **CSSStyleSheet** items.
And then, use the **adoptedStyleSheets** options, in order to style pages and components.
But the issue with these interfaces, is that they're not supported from Safari (and polyfill does not work great)

**Comment by alexander-akait** [comment]
@mikeaustin Can you prodvide example of the problem (small repo), so I provide configuration solutions how I will solve the same problems

**Comment by mditullio** [comment]
Hello @alexander-akait 
I did a small example of the issue: https://github.com/mditullio/demo-webpack-style-loader.
The need behind would be to share the same CSS from multiple web component instances.
Ideally, this could be, as suggested by @andriichumak a module which exports the URL / path to the extracted CSS.
I would see well a new inject type called `lazyLinkTag` which could something like that.
Thanks in advance for the attention!

**Comment by alexander-akait** [comment]
@mikeaustin Tha main problem style-loader is not extract CSS, it is for CSS in runtime code, anyway if you can use `mini-css-extract-plugin` and dynamic `import(...)` (alternative solution - set `runtime: false` and implement own logic for loading), but yes, `CSSStyleSheet` is right solution (that is bad it has a bad support right now, it was design for Shadow DOM). You have written polyfill is not wroking, can you desribe what is broken? Maybe you can create own small polyfill if you don't need extra complex logic

**Comment by mditullio** [comment]
Thanks for the suggestion about dynamic import, I will look onto it.

For the polyfill of adoptedStyleSheets, I tried this lib: https://github.com/calebdwilliams/construct-style-sheets
But there are some issues regarding Safari: https://github.com/calebdwilliams/construct-style-sheets/issues/110
But in general, to polyfill it, no other way than replicating style tags in each component instance.

So, for now, I decided to use Shadow DOM for intranet-only needs (just supporting Edge / Chrome), waiting for Apple updating its browsers... Safari is the new IE11 🐢

**Comment by alexander-akait** [comment]
Feel free to feedback and ask questions

**Comment by aw1875** [comment]
Just came across this issue as I'm having an issue with the same thing currently (I posted it on [StackOverflow](https://stackoverflow.com/questions/75190827/issues-with-style-loader-lazyloading-for-multiple-elements-in-shadow-dom) also). Reading through this I'm a little confused if someone found an appropriate workaround to this issue as I've been stuck on this for a few days now and I feel like I've tried a few different approaches.

**Comment by mditullio** [comment]
Hello @aw1875 

What you could do, if you can use **CSSStyleSheet** class (remark it's not available on every browser, Safari does not support it) is to directly export your stylesheet files in the form of **CSSStyleSheet** using the css-loader - look at https://webpack.js.org/loaders/css-loader/#exporttype.

The problem in my case, is that I needed to support also Safari browsers, so I ended up just by not using ShadowRoot, injecting them global styles using SASS with a root selector `name-of-my-component { ... }`.

**Comment by aw1875** [comment]
@mditullio appreciate the response! I looked through the docs and gave it a try based on them. It didn't seem to work, giving me the error "TypeError: Failed to set the 'adoptedStyleSheets' property on 'ShadowRoot': Failed to convert value to 'CSSStyleSheet'." I'm not sure if this is because I'm using tailwind and postcss but the documentation seemed pretty straight forward.

**Comment by mditullio** [comment]
Yes, style-loader should be removed if you use injectType = css-style-sheet.

**Comment by aw1875** [comment]
Just responded on StackOverflow but thank you it works perfectly now!

**Comment by mditullio** [comment]
Perfect ! You're welcome ;)
